pkg/scheduler/plugins/topology/topology_plugin_job_filtering.go

```diff
@@ -0,0 +1,277 @@
+// Copyright 2025 NVIDIA CORPORATION
+// SPDX-License-Identifier: Apache-2.0
+
+package topology
+
+import (
+	"fmt"
+	"slices"
+	"sort"
+
+	"github.com/NVIDIA/KAI-scheduler/pkg/scheduler/api/node_info"
+	"github.com/NVIDIA/KAI-scheduler/pkg/scheduler/api/pod_info"
+	"github.com/NVIDIA/KAI-scheduler/pkg/scheduler/api/podgroup_info"
+	"github.com/NVIDIA/KAI-scheduler/pkg/scheduler/api/resource_info"
+	"github.com/NVIDIA/KAI-scheduler/pkg/scheduler/log"
+)
+
+type jobAllocationMetaData struct {
+	maxPodResources    *resource_info.ResourceRequirements
+	allocationTestPods []*pod_info.PodInfo
+	tasksToAllocate    []*pod_info.PodInfo
+}
+
+func (t *topologyPlugin) prePredicateFn(_ *pod_info.PodInfo, job *podgroup_info.PodGroupInfo) error {
+	topologyTree, err := t.getJobTopology(job)
+	if err != nil {
+		return err
+	}
+	if topologyTree == nil {
+		return nil
+	}
+
+	// Calc tree job allocation data
+	maxAllocatablePods, err := t.calcTreeAllocatable(job, topologyTree)
+	if err != nil {
+		return err
+	}
+
+	// Get best domains for the job
+	var jobAllocateableDomain []*TopologyDomainInfo
+	if maxAllocatablePods >= len(podgroup_info.GetTasksToAllocate(job, t.taskOrderFunc, true)) {
+		jobAllocateableDomain, err = t.getBestjobAllocateableDomains(job, topologyTree)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Clean allocation data from the tree
+	for _, levelDomains := range topologyTree.DomainsByLevel {
```

Comment:
Make this a method of topologyTree (`topologyTree.Cleanup()` and call it with `defer` right after populating the values. Like closing a file

Reply:
Fixed